dashboard/lib/services/curriculum_pdfs.rb

```diff
@@ -1,22 +1,52 @@
 
-# Overview of all Curriculum PDFs
+# This module is the entry point for all "Curriculum PDFs", referring to the
+# PDFs that we automatically generate from our teacher-facing Lesson Plans and
+# related content.
 #
-# Current:
-# - teacher lesson plans (csp1-2021/20210216001309/teacher-lesson-plans/Welcome to CSP.pdf)
-# - student lesson plans (csp1-2021/20210216001309/student-lesson-plans/Welcome to CSP.pdf)
-# - script resources (csp1-2021/20210216001309/resources/CSP Unit 1 - Digital Information ('21-'22) - Resources.pdf)
-# - script overview (csp1-2021/20210216001309/CSP Unit 1 - Digital Information ('21-'22).pdf)
+# Note that, following Rails convention, this file defines the basic structure
+# of the module but most of the logic lives in the submodules, as defined in
+# the various other files in the `curriculum_pdfs/` directory.
+#
+# Currently (as of September 2021), we generate the following kind of PDFs:
+#
+# - Teacher-facing Lesson Plans (for example, https://lesson-plans.code.org/csp1-2021/20210826162223/teacher-lesson-plans/Welcome+to+CSP.pdf)
+# - Student-facing Lesson Plans (for example, https://lesson-plans.code.org/csp1-2021/20210826162223/student-lesson-plans/Welcome+to+CSP.pdf)
+# - Script Overviews (for example, https://lesson-plans.code.org/csp1-2021/20210826162223/Digital+Information+('21-'22).pdf)
+# - Script Resource Rollups (for example, https://lesson-plans.code.org/csp1-2021/20210826162223/Digital+Information+('21-'22)+-+Resources.pdf)
+#
+# We may also want in the future to generate more kinds, including:
 #
-# Future:
 # - student resources, based on Resource#audience
 # - unit calendar
 # - four rollup pages for unit, unit group, one for each of:
 #   - vocab
 #   - resouces
 #   - standards
 #   - programming expressions
+#
+# Our overall approach for PDF generation is as follows:
+#
+# When lesson plan is updated on levelbuilder, we serialize it into a
+# `.script_json` file with an `updated_at` timestamp. When that file is then
+# seeded on the staging machine, we compare that timestamp to the last known
+# value; if we detect that any content in the entire `.script_json` file has
+# been updated, we regenerate all PDFs for the associated script.
+#
+# PDFs are generated by puppeteer. Specifically, all content that is going to
+# be converted into a PDF is built with CSS print styles as well as web-view
+# styling, and we simply print the page to PDF with headless Chrome. Note that
+# this also means that anyone (including teachers) can generate their own PDFs
+# by simply printing the page from their own browser.
+#
+# After generating all PDFs on the staging machine during the staging build,
+# PDFs are uploaded to S3. They're labeled with the relevant `updated_at`
+# timestamp and served directly from S3, which allows us to generate the files
+# once as part of the build and then serve them from all of our various
+# environments without worrying about version mismatch. Note that this does
```

Comment:
@davidsbailey should we have a backlog item to make a plan for not storing pdfs indefinitely?

Reply:
my experience with S3 has been that it is so cheap that these are probably not worth cleaning up from a dollar cost perspective. since these are all sorted by timestamp, I'm not sure we'll have any problems with managing the complexity of storing so many versions either. so I would suggest no, let's not proactively track this. if scrolling through pages in the S3 UI ends up slowing us down in the future, then that could be our prompt to look into this further. How does that sound?

Reply:
Sounds good! Just wanted to check

Reply:
We're tracking this here https://codedotorg.atlassian.net/browse/PLAT-951